clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3602,8 +3602,12 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     BodyStmts.insert(BodyStmts.end(), FinalizeStmts.begin(),
                      FinalizeStmts.end());
 
+    const auto *LDcl = NewBody;
+    SourceLocation LL = LDcl ? LDcl->getBeginLoc() : SourceLocation();
+    SourceLocation LR = LDcl ? LDcl->getEndLoc() : SourceLocation();
+
     return CompoundStmt::Create(SemaSYCLRef.getASTContext(), BodyStmts,
-                                FPOptionsOverride(), {}, {});
+                                FPOptionsOverride(), LL, LR);
```

Comment:
What is this actually doing? This is supposed to correspond to location in source code IIUC which of course doesn't exist for compiler generated code. So what location is being picked up here?

Reply:
This is picking up the source location from the kernel body - so the compound statement uses the same location as the kernel body. You are right, this is a compiler-generated function, but it still needs a valid source location to ensure any tools processing the clang AST and requiring that source location to be valid to succeed. Other compiler-generated function like constructors also have valid locations even though they don't exist in source code.

Reply:
I think this sounds reasonable but I would like @tahonermann to also weigh in here. IIRC the work we re currently implementing for upstream clang (which will replace a lot of this code eventually) does not set SourceLocations as well.

Reply:
@elizabethandrews @tahonermann do you have updates on this? Thank you

Reply:
Sorry, I somehow missed @elizabethandrews's earlier request for me to weigh in on this.

I think this is reasonable. I'm not sure there is clear guidance for cases like this where a function definition is composed from both generated and written code.

There might be cases in the new upstream implementation that will need to have source locations added. I think I generally opted for generated code to have builtin/unknown source locations.